### PR TITLE
SNOW-165820 throw error when upload fails 

### PR DESCRIPTION
**Pull request by sfc-gh-vreddy**
Throws error when Put fails
Adds timestamps for put speeds

### CODE REVIEW

File: cpp/FileTransferAgent.cpp [hunk 1]
```diff
@@ -222,6 +223,11 @@ void Snowflake::Client::FileTransferAgent::upload(string *command)
         renewToken(command);
         i--;
       }
+      else if( outcome == RemoteStorageRequestOutcome::FAILED)
+      {
+        throw SnowflakeTransferException(TransferError::INTERNAL_ERROR,
```
Comment (sfc-gh-sshankar) [inline cpp/FileTransferAgent.cpp:228]: Can you improve this message ? What is the file we failed to upload, and why did it fail ?

Comment (sfc-gh-sshankar) [reply cpp/FileTransferAgent.cpp:228]: And what about reporting errors and checking for outcomes on small files ?

File: cpp/FileTransferAgent.cpp [hunk 2]
```diff
@@ -251,16 +257,23 @@ void Snowflake::Client::FileTransferAgent::uploadFilesInParallel(std::string *co
...
+            CXX_LOG_DEBUG("Token expired, Renewing token.")
             _mutex_lock(&m_parallelTokRenewMutex);
             this->renewToken(command);
             _mutex_unlock(&m_parallelTokRenewMutex);
           }
           else
           {
+            if( outcome == RemoteStorageRequestOutcome::FAILED)
+            {
+              throw SnowflakeTransferException(TransferError::INTERNAL_ERROR,
```
Comment (sfc-gh-sshankar) [inline cpp/FileTransferAgent.cpp:274]: Again, add more information to the user-visible error message

File: cpp/SnowflakeAzureClient.cpp [hunk 3]
```diff
@@ -49,14 +49,16 @@ SnowflakeAzureClient::SnowflakeAzureClient(StageInfo *stageInfo,
       snowflake_global_get_attribute(SF_GLOBAL_CA_BUNDLE_FILE, caBundleFile, sizeof(caBundleFile));
       CXX_LOG_TRACE("ca bundle file from SF_GLOBAL_CA_BUNDLE_FILE *%s*", caBundleFile);
   }
-if( caBundleFile[0] == 0 ) {
+  else if( caBundleFile[0] == 0 ) {
```
Comment (sfc-gh-sshankar) [inline cpp/SnowflakeAzureClient.cpp:52]: Why the else if, don't you always want to check this.

Comment (sfc-gh-vreddy) [reply cpp/SnowflakeAzureClient.cpp:52]: Thats true, The else in line 48 also needs to be removed.

File: cpp/FileTransferAgent.cpp (hunk 1 above)
Comment (sfc-gh-vreddy) [reply cpp/FileTransferAgent.cpp:228]: @sfc-gh-sshankar For small files its done on line 272.
We did add at multiple locations the reason for error. 
Will add more detail to the exception.

File: cpp/FileTransferAgent.cpp (hunk 2 above)
Comment (sfc-gh-vreddy) [reply cpp/FileTransferAgent.cpp:274]: Yes, will add the reason for failure and the file name.

**Comment by sfc-gh-sshankar** [comment]
How are we testing whether an error is thrown when a put fails ?

**Comment by sfc-gh-stakeda** [comment]
Would you add tests in libsnowflakeclient?

**Comment by sfc-gh-vreddy** [comment]
@smtakeda Added integration tests here to validate different FILE URIS.

**Comment by sfc-gh-vreddy** [comment]
These tests fail on docker. Looking into why its failing in docker.

**Pull request merged**